### PR TITLE
Chore switch to postgres db

### PR DESCRIPTION
**Pull request by Cheza-Dzabala**
#### What does this PR do?

Removes all data structures from the application & implements a Postgresql Database.

#### Description of Task to be completed?

Removing all data structures from the application.  Replaced the datas tructures with 3 data bases:
 
Users
 ------------------ 
**id** | SERIAL INTEGER PRIMARY KEY
**first_name** | VARCHAR (128) NOT NULL
**last_name** | VARCHAR (128) NOT NULL
**email** | VARCHAR (128) NOT NULL
**password** | VARCHAR (128) NOT NULL

 Trips
 ------------------ 
**id** | SERIAL INTEGER PRIMARY KEY
**trip_date** | VARCHAR (128) NOT NULL
**bus_licesnse_number** | VARCHAR (128) NOT NULL
**origin** | VARCHAR (128) NOT NULL
**destination** | VARCHAR (128) NOT NULL
**fare** | VARCHAR (128) NOT NULL
**seating_capacity** | INTEGER NOT NULL
**number_of_seats** | INTEGENOT NULL

 Bookings
 ------------------ 
**id** | SERIAL INTEGER
**seat_number** | VARCHAR (128) NOT NULL
**user_id** | VARCHAR (128) NOT NULL
**trip_id** | VARCHAR (128) NOT NULL
PRIMARY KEY (user_id, trip_id)

#### How should this be manually tested?

1. Install PostgreSQL by following [these](https://www.postgresql.org/docs/9.3/tutorial-install.html) instructions.
2. Start PostgreSQL. 
3. Clone the repo. 
4. Install all dependencies by running ` npm install `.
5. Run the command: ` npm start `
6. Assuming you're on localhost:3000, sign up for account at ` localhost:3000/api/v2/auth/signup `
7. Check PostgreSQL for persistance. 
<-- opttional -->

To access protected routes:

1. Default user email: ` dzabalamacheza@gmail.com ` & password ` Runfree8418_!* `
2. To create your own admin, access route:
POST ` localhost:3000/api/v2/admins `

###### Entity Specification
```javascript
body: {
  "email": String,
   "first_name": String,
    "last_name": String,
    "password": String
}
````

###### Response Specification
```javascript
body: {
  "status": "success",
 "id": Int,
   "first_name": String,
    "last_name": String,
}
````

#### Any background context you want to provide?

Data was previously stored in Arrays

#### Screenshots (if appropriate)

Screenshot 2019-08-14 at 9 09 09 AM

screencapture-127-0-0-1-49164-browser-2019-08-14-09_09_20

Screenshot 2019-08-14 at 9 08 32 AM

Screenshot 2019-08-14 at 9 18 17 AM

Screenshot 2019-08-14 at 9 07 41 AM

screencapture-127-0-0-1-49164-browser-2019-08-14-09_07_49

Updated Codebase
-------------------

Screenshot 2019-08-14 at 7 48 50 PM
Screenshot 2019-08-14 at 7 48 37 PM
Screenshot 2019-08-14 at 7 48 18 PM
Screenshot 2019-08-14 at 7 48 06 PM
Screenshot 2019-08-14 at 7 47 53 PM
Screenshot 2019-08-14 at 7 47 34 PM
Screenshot 2019-08-14 at 7 47 24 PM

#### Questions:

None

### CODE REVIEW

File: server/v2/app.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+app.use(morgan('dev'));
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+
+
+app.use(`/api/${version}/auth`, authRoutes);
+app.use(`/api/${version}/trips`, tripRoutes);
+app.use(`/api/${version}/bookings`, bookingRoutes);
+app.use(`/api/${version}/admins`, adminRoutes);
```
Comment (Oluwasegun-AA) [inline server/v2/app.js]: @Cheza-Dzabala , with respect to the directions IU gave regarding making your application modular, you should only use a single middleware for your custom routes.  this should be imported from the index.js file in the middleware folder

File: server/v2/helpers/bookingHelpers/bookingRelationships.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  const authenticated = typeCheck.getUser(req);
+  const user = await userModel.findUserById(authenticated.id);
+  const trip = await tripsModel.findTrip(req.body.trip_id);
+  if (!user) return Response(res, 400, 'unsuccessful', { message: 'No user found with this ID' });
+  if (!trip) return Response(res, 400, 'unsuccessful', { message: 'No trip found with this ID' });
+  req.body.user_id = user.id;
+  req.body.trip_id = trip.id;
+  req.body.email = user.email;
+  return next();
+};
```
Comment (Oluwasegun-AA) [inline server/v2/helpers/bookingHelpers/bookingRelationships.js]: @Cheza-Dzabala , this is a middleware, it should not be in the helpers folder; rather this should be in the middleware folder

File: server/v2/helpers/bookingHelpers/bookingSeatChecker.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    if (booking.allocated_seat === body.seat_number && booking.trip_id === body.trip_id) {
+      flag = { status: true, message: 'This seat is unavailable' };
+    } else if (booking.user_email === body.email && booking.trip_id === body.trip_id) {
+      flag = { status: true, message: 'You\'ve already made a booking on this bus' };
+    }
+  });
+  if (flag.status === true) return Response(res, 409, 'unsuccessful', { message: flag.message });
+
+  return next();
+};
```
Comment (Oluwasegun-AA) [inline server/v2/helpers/bookingHelpers/bookingSeatChecker.js]: @Cheza-Dzabala , middlwares should not be in the helpers folder

File: server/v2/helpers/bookingHelpers/verifyBooking.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import bookingsModel from '../../models/bookingsModel';
+import Response from '../responseHelper';
+
+export default async (req, res, next) => {
+  const { id } = req.params;
+  const booking = await bookingsModel.findBooking(id);
+  if (!booking) return Response(res, 404, 'unsuccessful', { message: 'Booking already deleted' });
+  return next();
+};
```
Comment (Oluwasegun-AA) [inline server/v2/helpers/bookingHelpers/verifyBooking.js]: @Cheza-Dzabala , i can see you have so many more middlewares in the helpers folder. please remove them and structure your files accordingly

File: server/v2/helpers/tripHelpers/tripsRelationships.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        if (busses[len].trip_date === body.trip_date) {
+          return Response(res, 209, 'unsuccessful', { message: 'This bus is already booked on a trip on this date' });
+        }
+      }
+      len += 1;
+    }
+  }
+
+  return next();
+};
```
Comment (Oluwasegun-AA) [inline server/v2/helpers/tripHelpers/tripsRelationships.js]: @Cheza-Dzabala please explain in this conversation what you are trying to achieve here

File: server/v2/middleware/userTypeMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable dot-notation */
```
Comment (Oluwasegun-AA) [inline server/v2/middleware/userTypeMiddleware.js]: @Cheza-Dzabala , you are required the follow strictly the Airbnb Eslint configuration. in lieu of this, you are not expected to bypass an enforced style using the Esling comments. please remove all comments and ensure your code conforms to the Airbnb standard without errors

File: server/v2/middleware/validationMiddleware/signupValidationMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import userModel from '../../models/userModel';
+
+import authValidations from '../../validations/authValidations';
+
```
Comment (Oluwasegun-AA) [inline server/v2/middleware/validationMiddleware/signupValidationMiddleware.js]: @Cheza-Dzabala , you might want to remove unnecessary spaces between your import statements

File: server/v2/models/userModel.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+const hashIt = password => bcrypt.hashSync(password, bcrypt.genSaltSync(8), null);
+async function findUser(email) {
+  const user = await dbService.findUser(email);
+  return user;
+}
+
+async function findUserById(id) {
+  const user = await dbService.findUserById(id);
+  return user;
+}
```
Comment (Oluwasegun-AA) [inline server/v2/models/userModel.js:36]: @Cheza-Dzabala , you have adopted a nice apprach. however, I can see you'r repeating methods. why can you have a single class that will handle all kinds of request, whether user/booking or trip?

above there you had a method to fetch `all trips/trips by Id` and another method to fetch `all booking/ booking by Id`. you can basically have just a single method(in the helper file) which implement this `ONCE` for all the tables. hence you call it in all controllers irrespective

File: server/v2/package.json [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    "swagger-ui-express": "^4.0.7"
+  },
+  "devDependencies": {
+    "coveralls": "^3.0.5",
+    "eslint": "^5.16.0",
+    "eslint-config-airbnb-base": "^13.2.0",
+    "eslint-plugin-import": "^2.18.2",
+    "mocha": "^6.2.0",
+    "mocha-lcov-reporter": "^1.3.0",
+    "nyc": "^14.1.1"
```
Comment (Oluwasegun-AA) [inline server/v2/package.json]: @Cheza-Dzabala , please remove all the caret (^)symbols in your dependencies and dev-dependencies. it ensures your files are locked and breaking changes in future version will not affect your workflow even when you rerun this app in the year 2050

File: server/v2/services/authDBServices.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  try {
+    const { rows } = await query(queryText, values);
+    return rows.length;
+  } catch (error) {
+    console.log(`Database ${error}`);
+  }
+}
+
+async function signin(email, password) {
+  const queryText = 'SELECT * From users WHERE email = $1';
```
Comment (Oluwasegun-AA) [inline server/v2/services/authDBServices.js]: @Cheza-Dzabala , for modularity sake, you can have a helper file which is going to house all your query texts.. so you do somthing like 
```js
import * as query from './queryHelperFile'

const queryText = query.getBookings;

```

File: server/v2/controllers/signinController.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import authenticationCheck from '../middleware/authenticationCheck';
+import UserModel from '../models/userModel';
+
+import Response from '../helpers/responseHelper';
+
+async function signin(body, res) {
+  const { email, password } = body;
+  const user = await UserModel.signin(email, password);
+  if (user) {
+    return authenticationCheck.signNewToken(user, res, 200);
```
Comment (Oluwasegun-AA) [inline server/v2/controllers/signinController.js]: @Cheza-Dzabala , are you encrypting all user data here? if yes, that is a bad practice. you should only encrypt few data that will be useful on the front end of things (i.e for making a return request)

**Review by Oluwasegun-AA — CHANGES_REQUESTED** [review]

File: server/v2/app.js (hunk 1 above)
Comment (Cheza-Dzabala) [reply server/v2/app.js]: Well noted.

File: server/v2/helpers/bookingHelpers/bookingRelationships.js (hunk 2 above)
Comment (Cheza-Dzabala) [reply server/v2/helpers/bookingHelpers/bookingRelationships.js]: Noted.. Will resolve this immidiately.

File: server/v2/helpers/bookingHelpers/bookingSeatChecker.js (hunk 3 above)
Comment (Cheza-Dzabala) [reply server/v2/helpers/bookingHelpers/bookingSeatChecker.js]: Noted

File: server/v2/helpers/bookingHelpers/verifyBooking.js (hunk 4 above)
Comment (Cheza-Dzabala) [reply server/v2/helpers/bookingHelpers/verifyBooking.js]: Noted.

File: server/v2/helpers/tripHelpers/tripsRelationships.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import tripModel from '../../models/tripsModel';
+import Response from '../responseHelper';
+
+export default async (req, res, next) => {
+  const { body } = req;
+  const busses = await tripModel.findBus(body.bus_license_number);
+  let len = 0;
+  if (busses.length > 0) {
```
Comment (Cheza-Dzabala) [inline server/v2/helpers/tripHelpers/tripsRelationships.js]: Here I determine if busses have been retrieved from the database

File: server/v2/helpers/tripHelpers/tripsRelationships.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    while (len < busses.length) {
```
Comment (Cheza-Dzabala) [inline server/v2/helpers/tripHelpers/tripsRelationships.js]: I loop through the busses that were retrieved from the database to determine if they have the same trip date as the trip date in the request body. If the dates are the same, Then I determine that the bus is booked.

File: server/v2/middleware/userTypeMiddleware.js (hunk 6 above)
Comment (Cheza-Dzabala) [reply server/v2/middleware/userTypeMiddleware.js]: Yes, I found the red underlining to be annoying but I will work with it and refactor to the recommended language

File: server/v2/middleware/validationMiddleware/signupValidationMiddleware.js (hunk 7 above)
Comment (Cheza-Dzabala) [reply server/v2/middleware/validationMiddleware/signupValidationMiddleware.js]: WIll do that

File: server/v2/models/userModel.js (hunk 8 above)
Comment (Cheza-Dzabala) [reply server/v2/models/userModel.js:36]: This is an interesting suggestion. I will adopt it.

File: server/v2/package.json (hunk 9 above)
Comment (Cheza-Dzabala) [reply server/v2/package.json]: Noted

File: server/v2/services/authDBServices.js (hunk 10 above)
Comment (Cheza-Dzabala) [reply server/v2/services/authDBServices.js]: Ah yes, I will do that.

File: server/v2/controllers/signinController.js (hunk 11 above)
Comment (Cheza-Dzabala) [reply server/v2/controllers/signinController.js]: I have a middleware that encrypts the data. I am sending that middleware the user object, the response and the appropriate status code to return on success.

**Pull request merged**